### ISSUE TITLE
Give middleware access to projection

### ISSUE THREAD

**Issue by Anthonyzou** [graphql-compose/graphql-compose#170]
I am trying to update my codebase to use middleware but it lacks the ability to manipulate the projection of resolve params. Is there a way to do this?

**Comment by nodkz** [comment]
For now in Prisma middlewares now such property. And I do not want to change the shape of middleware for keeping compatibility.

For now you may get `projection` in such way:

```js
import { getProjectionFromAST, getFlatProjectionFromAST } from 'graphql-compose';

export async function middleware(resolve, source, args, context, info) {
  console.log(getProjectionFromAST(info));
  console.log(getFlatProjectionFromAST(info));
}
```

**Comment by Anthonyzou** [comment]
I see, is there a way to change the projection back into the info AST?

**Comment by nodkz** [comment]
Nope. If you need such method then you do something wrong. Try to find another solution

**Comment by toverux** [comment]
I'm also migrating to middlewares, and I'm also stuck with this :)
I also need `projection`, but most importantly, `beforeRecordMutate`.
The generic resolve wrapper that made use of `beforeRecordMutate` served as a helper that checked the users' right to mutate documents.
I think I'm gonna roll out my own `withMiddlewares` function for now so I can continue to port my old code. Otherwise, I don't see any alternative other than coming back to my old custom syntax and helpers. I wanted to remove that code now that graphql-compose offered an ootb solution, but that won't be quite enough for now.
What is your concern with exposing the full object just to middlewares ?

**Comment by nodkz** [comment]
@toverux, for now, your solution is right. 

`beforeRecordMutate` is something very specific to `graphql-compose-mongoose` and not applicable to other plugins. I want to keep `graphql-compose` as much independent from plugins as possible. So `graphql-compose` should not know anything about `beforeRecordMutate`.

Maybe we can somehow improve graphql-compose-mongoose resolvers to fit your requirements? Conver them to some configurable classes or something other.

For now, it's not enough information for me about your problem, to suggest some great solution.
